### PR TITLE
Monte Carlo time stepping action

### PR DESCRIPTION
**Pull request by ffoucart**
## Proposed changes

Add an action to take a single MC time step, and a test of using this action within the mock run environment.
Only the last commit is actually relevant; all others are from PR #6138 , which is a prerequisite to this PR.

### Upgrade instructions

None

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### CODE REVIEW

File: src/Evolution/Particles/MonteCarlo/Actions/TimeStepActions.hpp [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+#include "Parallel/AlgorithmExecution.hpp"
+#include "Parallel/GlobalCache.hpp"
+#include "PointwiseFunctions/GeneralRelativity/Tags.hpp"
+#include "PointwiseFunctions/Hydro/EquationsOfState/EquationOfState.hpp"
+#include "PointwiseFunctions/Hydro/Tags.hpp"
+#include "Time/Tags/TimeStepId.hpp"
+#include "Time/Time.hpp"
+#include "Time/TimeStepId.hpp"
+
+#include "Parallel/Printf/Printf.hpp"
```
Comment (nilsdeppe) [inline src/Evolution/Particles/MonteCarlo/Actions/TimeStepActions.hpp]: remove? Leftover from debugging?

File: src/Evolution/Particles/MonteCarlo/Actions/TimeStepActions.hpp [hunk 2]
```diff
@@ -0,0 +1,168 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
```
Comment (nilsdeppe) [inline src/Evolution/Particles/MonteCarlo/Actions/TimeStepActions.hpp:5]: Need:
```cpp
#include <array>
#include <cstddef>
#include <optional>
#include <random>
#include <vector>
```

File: src/Evolution/Particles/MonteCarlo/Actions/TimeStepActions.hpp [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+#include "PointwiseFunctions/Hydro/Tags.hpp"
+#include "Time/Tags/TimeStepId.hpp"
+#include "Time/Time.hpp"
+#include "Time/TimeStepId.hpp"
+
+#include "Parallel/Printf/Printf.hpp"
+
+namespace Particles::MonteCarlo {
+
+// Mutator advancing neutrinos by a single step
```
Comment (nilsdeppe) [inline src/Evolution/Particles/MonteCarlo/Actions/TimeStepActions.hpp]: triple slash for dox comment

File: src/Evolution/Particles/MonteCarlo/Actions/TimeStepActions.hpp [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+      evolution::dg::subcell::fd::Tags::DetInverseJacobianLogicalToInertial,
+      domain::Tags::InverseJacobian<Dim + 1, Frame::Inertial, Frame::Fluid>,
+      domain::Tags::Jacobian<Dim + 1, Frame::Inertial, Frame::Fluid>,
+      Particles::MonteCarlo::Tags::MortarDataTag<Dim>>;
+
+  static void apply(
+      gsl::not_null<std::vector<Packet>*> packets,
+      gsl::not_null<std::mt19937*> random_number_generator,
+      gsl::not_null<std::array<DataVector, NeutrinoSpecies>*>
+          single_packet_energy,
```
Comment (nilsdeppe) [inline src/Evolution/Particles/MonteCarlo/Actions/TimeStepActions.hpp]: `const gsl::not_null`

File: src/Evolution/Particles/MonteCarlo/Actions/TimeStepActions.hpp [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+        inertial_to_fluid_jacobian, inertial_to_fluid_inverse_jacobian,
+        electron_fraction_ghost, baryon_density_ghost, temperature_ghost,
+        cell_light_crossing_time_ghost);
+  }
+};
+
+namespace Actions {
+
+template <size_t EnergyBins, size_t NeutrinoSpecies>
+struct TakeTimeStep {
```
Comment (nilsdeppe) [inline src/Evolution/Particles/MonteCarlo/Actions/TimeStepActions.hpp:157]: please add short dox comment :)

File: src/Evolution/Particles/MonteCarlo/Actions/TimeStepActions.hpp [hunk 6]
```diff
@@ -0,0 +1,168 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include "DataStructures/DataBox/DataBox.hpp"
+#include "Domain/Tags.hpp"
+#include "Domain/TagsTimeDependent.hpp"
+#include "Evolution/DgSubcell/ActiveGrid.hpp"
+#include "Evolution/DgSubcell/Tags/Coordinates.hpp"
```
Comment (nilsdeppe) [inline src/Evolution/Particles/MonteCarlo/Actions/TimeStepActions.hpp:17]: Add include:
```cpp
#include "Evolution/DgSubcell/Tags/ActiveGrid.hpp"
```

File: src/Evolution/Particles/MonteCarlo/NeutrinoInteractionTable.hpp [hunk 7]
```diff
@@ -34,6 +36,12 @@ class NeutrinoInteractionTable {
       std::vector<double> table_log_temperature_,
       std::vector<double> table_electron_fraction_);
 
+  explicit NeutrinoInteractionTable(CkMigrateMessage* msg) : PUP::able(msg) {}
+
+  using PUP::able::register_constructor;
+  void pup(PUP::er& p);
```
Comment (nilsdeppe) [inline src/Evolution/Particles/MonteCarlo/NeutrinoInteractionTable.hpp]:
```cpp
void pup(PUP::er& p) override;
```

File: src/Evolution/Particles/MonteCarlo/NeutrinoInteractionTable.cpp [hunk 8]
```diff
@@ -245,6 +245,23 @@ NeutrinoInteractionTable<EnergyBins, NeutrinoSpecies>::NeutrinoInteractionTable(
   initialize_interpolator();
 }
 
+template <size_t EnergyBins, size_t NeutrinoSpecies>
+void NeutrinoInteractionTable<EnergyBins, NeutrinoSpecies>::pup(PUP::er& p) {
+  p | table_data;
```
Comment (nilsdeppe) [inline src/Evolution/Particles/MonteCarlo/NeutrinoInteractionTable.cpp:251]: Please add as first line:
```cpp
PUP::able::pup(p);
```
I think we need that for charm

File: src/Evolution/Particles/MonteCarlo/Tags.hpp [hunk 9]
```diff
@@ -23,4 +25,19 @@ struct CellLightCrossingTime : db::SimpleTag {
   using type = Scalar<DataType>;
 };
 
+struct RandomNumberGenerator : db::SimpleTag {
```
Comment (nilsdeppe) [inline src/Evolution/Particles/MonteCarlo/Tags.hpp:35]: dox 3x

**Review by nilsdeppe — CHANGES_REQUESTED** [review]
LGTM, a few suggestions. You can squash them directly

**Comment by ffoucart** [comment]
Ok, I had missed a few initially, but I think now all comments have been addressed...

**Review by nilsdeppe — APPROVED** [review]

**Comment by nilsdeppe** [comment]
CI is timeout in Cylindrical BBH domain, which is unrelated.

**Pull request merged**